### PR TITLE
[.NET] Inline Text Runs

### LINKED ISSUE #2378: [.NET] Inline text runs
**Automatically generated task. Any updates to the title and body will be overwritten. Do NOT modify the labels.**

.NET implementation of #1933

### PR DESCRIPTION
**Pull request by carlos-zamora**
Closes #2378 (Spec #1933 )

# Serialization
* Created `ITextElement` interface to make handle overlap between `TextRun` and `TextBlock`.
* `RichTextBlock` introduced as new element
  * encloses `AdaptiveParagraph` which encloses `AdaptiveInlineRun`
* `AdaptiveTextRun` introduced as `AdaptiveInlineRun` adhering to `ITextElement` interface
* `RichTextBlock` (De)Serialization tested

# HTML
Mostly straightforward. `Microsoft.MarkedNet` wraps processed text in `<p>` tag for whatever reason so that tag has to get removed each time an `InlineTextRun` is rendered.

# WPF
Also straightforward.

### CODE REVIEW

**Comment by carlos-zamora** [comment]
@matthidinger @andrewleader The `MaxLines` functionality is a little weird. Basically, since inline text runs can be of different sizes, I make `MaxLines` apply based on the **largest** font used. I wanted to run it by you guys if this was fine or if I should rewrite this behavior. Thoughts?

**Comment by paulcam206** [comment]
.net build has test failures

```
Failed   SerializeAllScenarios
Error Message:
 Assert.IsTrue failed. 
Begin Differences (1 differences):
Types [String,String], Item Expected.Body[3].Columns[1].Items[0].Text != Actual.Body[3].Columns[1].Items[0].Text, Values ( ,)
End Differences (Maximum of 1 differences shown).
Stack Trace:
   at AdaptiveCards.Test.XmlSerializationTests.SerializeAllScenarios() in C:\BA\1917\s\source\dotnet\test\AdaptiveCards.Test\XmlSerializationTests.cs:line 48
```

File: source/dotnet/Library/AdaptiveCards/AdaptiveRichTextBlock.cs [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+
+                /// <summary>
+                ///     Make the text less prominent
+                /// </summary>
+                public bool IsSubtle { get; set; }
+
+                /// <summary>
+                ///     The text to display
+                /// </summary>
+                public string Text { get; set; } = " ";
```
Comment (paulcam206) [inline source/dotnet/Library/AdaptiveCards/AdaptiveRichTextBlock.cs]: >  [](start = 52, length = 1)

is this your test failure?

File: source/dotnet/Library/AdaptiveCards/AdaptiveRichTextBlock.cs [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+        [JsonObject(NamingStrategyType = typeof(CamelCaseNamingStrategy))]
+        public class AdaptiveParagraph
+        {
+            [JsonConverter(typeof(IgnoreEmptyItemsConverter<AdaptiveInlineRun>))]
+#if !NETSTANDARD1_3
+            [XmlElement(typeof(AdaptiveTextRun))]
+#endif
+            public List<AdaptiveInlineRun> Inlines { get; set; } = new List<AdaptiveInlineRun>();
+
+            public abstract class AdaptiveInlineRun : AdaptiveTypedElement { }
```
Comment (paulcam206) [inline source/dotnet/Library/AdaptiveCards/AdaptiveRichTextBlock.cs]: >public abstract class AdaptiveInlineRun : AdaptiveTypedElement { } [](start = 12, length = 66)

let's skip this for now

File: source/dotnet/Test/AdaptiveCards.Test/SerializationTests.cs [hunk 3]
```diff
@@ -496,6 +496,130 @@ public void BackgroundImage()
...
+
+            richTB.Paragraphs.Add(paragraph1);
+
+            // Build Second Paragraph (Empty inlines)
+            var paragraph2 = new AdaptiveRichTextBlock.AdaptiveParagraph();
+            richTB.Paragraphs.Add(paragraph2);
+
+            card.Body.Add(richTB);
+
+            var expected = @"{
```
Comment (paulcam206) [inline source/dotnet/Test/AdaptiveCards.Test/SerializationTests.cs]: >var expected = @"{ [](start = 12, length = 18)

as noted in person, I think we would benefit from sample cards. be clear about when missing inlines or paragraphs become an error. also, show off markdown in inline runs! :)

**Comment by paulcam206** [comment]
> 

not your fault, but while you're here, please fix up these tabs

---
Refers to: source/dotnet/Library/AdaptiveCards.Rendering.Wpf/Helpers/ControlExtensions.cs:87 in 2e2b309. [](commit_id = 2e2b3099254cb85810aa0f607eb2ea69bcc27748, deletion_comment = False)

File: source/dotnet/Library/AdaptiveCards.Rendering.Html/AdaptiveCardRenderer.cs [hunk 4]
```diff
@@ -839,6 +842,129 @@ protected static HtmlTag TextBlockRender(AdaptiveTextBlock textBlock, AdaptiveRe
...
+                uiTextBlock.Style("flex", "1 1 100%");
+            }
+
+            if (!richTextBlock.IsVisible)
+            {
+                uiTextBlock.Style("display", "none");
+            }
+
+            var setWrapStyleOnParagraph = false;
+            if (richTextBlock.Wrap == false)
```
Comment (paulcam206) [inline source/dotnet/Library/AdaptiveCards.Rendering.Html/AdaptiveCardRenderer.cs]: this feels a bit awkward

File: source/dotnet/Library/AdaptiveCards.Rendering.Wpf/AdaptiveRichTextBlockRenderer.cs [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+                    colorOption = context.ForegroundColors.Light;
+                    break;
+                case AdaptiveTextColor.Warning:
+                    colorOption = context.ForegroundColors.Warning;
+                    break;
+                case AdaptiveTextColor.Default:
+                default:
+                    colorOption = context.ForegroundColors.Default;
+                    break;
+            }
```
Comment (paulcam206) [inline source/dotnet/Library/AdaptiveCards.Rendering.Wpf/AdaptiveRichTextBlockRenderer.cs]: is there not already a helper for this?

File: source/dotnet/Library/AdaptiveCards.Rendering.Wpf/AdaptiveRichTextBlockRenderer.cs [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+        }
+
+        private static FontWeight SelectMaxTextWeight(FontWeight first, FontWeight second)
+        {
+            return (first > second) ? first : second;
+        }
+
+        private static double SelectMaxTextSize(double first, double second)
+        {
+            return (first > second) ? first : second;
```
Comment (paulcam206) [inline source/dotnet/Library/AdaptiveCards.Rendering.Wpf/AdaptiveRichTextBlockRenderer.cs]: `Math.Max()`?

**Comment by andrewleader** [comment]
@carlos-zamora wow I'm really surprised that WPF TextBlock doesn't have a MaxLines property... but it's true that it doesn't! People actually have to work around that like how you did.

I think what you implemented sounds good enough, considering it seems there's no real solution to this.

File: source/dotnet/Library/AdaptiveCards/AdaptiveRichTextBlock.cs (hunk 1 above)
Comment (carlos-zamora) [reply source/dotnet/Library/AdaptiveCards/AdaptiveRichTextBlock.cs]: Nope. :( Besides, this is the same default value as for regular text block so I'll keep it around anyways.

---
In reply to: [263149504](https://github.com/Microsoft/AdaptiveCards/pull/2519#discussion_r263149504) [](ancestors = 263149504)

File: source/dotnet/Library/AdaptiveCards.Rendering.Html/AdaptiveCardRenderer.cs (hunk 4 above)
Comment (carlos-zamora) [reply source/dotnet/Library/AdaptiveCards.Rendering.Html/AdaptiveCardRenderer.cs]: This was all copied over from the TextBlockRenderer. Just replaced it with the ternary operator and got rid of the extra boolean in both.

---
In reply to: [263163649](https://github.com/Microsoft/AdaptiveCards/pull/2519#discussion_r263163649) [](ancestors = 263163649)

File: source/dotnet/Library/AdaptiveCards.Rendering.Wpf/AdaptiveRichTextBlockRenderer.cs (hunk 5 above)
Comment (carlos-zamora) [reply source/dotnet/Library/AdaptiveCards.Rendering.Wpf/AdaptiveRichTextBlockRenderer.cs]: Huh. Overlooked making a helper for WPF. Just added a helper for it.

---
In reply to: [263170355](https://github.com/Microsoft/AdaptiveCards/pull/2519#discussion_r263170355) [](ancestors = 263170355)

**Comment by paulcam206** [comment]
> 

thanks!

---
In reply to: [470301138](https://github.com/Microsoft/AdaptiveCards/pull/2519#issuecomment-470301138) [](ancestors = 470301138)

---
Refers to: source/dotnet/Library/AdaptiveCards.Rendering.Wpf/Helpers/ControlExtensions.cs:87 in 2e2b309. [](commit_id = 2e2b3099254cb85810aa0f607eb2ea69bcc27748, deletion_comment = False)

File: source/uwp/Renderer/lib/AdaptiveBackgroundImage.cpp [hunk 7]
```diff
@@ -25,12 +25,16 @@ namespace AdaptiveNamespace
...
-        if (sharedImage == nullptr || sharedImage->GetUrl().empty())
+        if (sharedImage == nullptr)
         {
             return E_INVALIDARG;
         }
 
-        RETURN_IF_FAILED(UTF8ToHString(sharedImage->GetUrl(), m_url.GetAddressOf()));
+        if (!sharedImage->GetUrl().empty())
+        {
+            RETURN_IF_FAILED(UTF8ToHString(sharedImage->GetUrl(), m_url.GetAddressOf()));
```
Comment (paulcam206) [inline source/uwp/Renderer/lib/AdaptiveBackgroundImage.cpp]: >RETURN_IF_FAILED [](start = 12, length = 16)

not sure this change should be part of your commit :)

File: source/uwp/UWPUnitTests/ObjectModelTests.cs [hunk 8]
```diff
@@ -26,7 +26,7 @@ public void Card()
                 VerticalContentAlignment = VerticalContentAlignment.Center,
             };
 
-            Assert.AreEqual("https://www.stuff.com/background.jpg", card.BackgroundImage);
+            Assert.AreEqual("https://www.stuff.com/background.jpg", card.BackgroundImage.Url);
```
Comment (paulcam206) [inline source/uwp/UWPUnitTests/ObjectModelTests.cs]: >kgroundImage. [](start = 76, length = 13)

not sure this change should be part of your commit :)

File: source/dotnet/Library/AdaptiveCards/AdaptiveRichTextBlock.cs [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+            [JsonConverter(typeof(IgnoreEmptyItemsConverter<AdaptiveTextRun>))]
+#if !NETSTANDARD1_3
+            [XmlElement(typeof(AdaptiveTextRun))]
+#endif
+            public List<AdaptiveTextRun> Inlines { get; set; } = new List<AdaptiveTextRun>();
+
+#if !NETSTANDARD1_3
+            [XmlType(TypeName = AdaptiveTextBlock.TypeName)]
+#endif
+            public class AdaptiveTextRun : AdaptiveTypedElement, ITextElement
```
Comment (RebeccaAnne) [inline source/dotnet/Library/AdaptiveCards/AdaptiveRichTextBlock.cs]: >AdaptiveTextRun [](start = 25, length = 15)

Looks like we need SelectAction on the runs as well

**Pull request merged**